experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/schemaBuilder.ts

```diff
@@ -0,0 +1,127 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Assume, RestrictiveReadonlyRecord, transformObjectMap } from "../../../util";
+import { typeNameSymbol } from "../../contextuallyTyped";
+import { FieldKind } from "../../modular-schema";
+import { ImplicitFieldSchema, NormalizeField, SchemaBuilderBase } from "../../schemaBuilderBase";
+import {
+	FieldSchema,
+	InternalTypedSchemaTypes,
+	StructSchema,
+	TreeNodeSchema,
+	Unenforced,
+} from "../../typed-schema";
+import { ProxyNode, SharedTreeObject } from "./types";
+
+const factoryContent = Symbol("Node content");
+interface HasFactoryContent<T> {
+	[factoryContent]: T;
+}
+
+/**
+ * Returns the content stored on an object created by a {@link SharedTreeObjectFactory}.
+ */
+export function getFactoryContent(x: unknown): unknown | undefined {
+	return (x as Partial<HasFactoryContent<unknown>>)[factoryContent];
+}
+
+/**
+ * Creates `{@link SharedTreeObject}`s of some type via a `create` method.
+ * @alpha
+ */
+export interface SharedTreeObjectFactory<TSchema extends TreeNodeSchema<any, any>> {
+	/**
+	 * Create a {@link SharedTreeObject} that can be inserted into the tree via assignment `=`.
+	 * @param content - the data making up the {@link SharedTreeObject} to be created.
+	 * @remarks
+	 * The {@link SharedTreeObject} created by this function may _only_ be used for insertion into the tree.
+	 * It may not be read, mutated or queried in any way.
+	 */
+	create(
+		content: ProxyNode<Assume<TSchema, StructSchema>, "javaScript">,
+	): SharedTreeObject<Assume<TSchema, StructSchema>>;
+}
+
+class FactoryTreeNodeSchema<
+		Name extends string = string,
+		T extends
+			Unenforced<InternalTypedSchemaTypes.TreeSchemaSpecification> = InternalTypedSchemaTypes.TreeSchemaSpecification,
+	>
+	extends TreeNodeSchema<Name, T>
+	implements SharedTreeObjectFactory<TreeNodeSchema<Name, T>>
+{
+	public create(
+		content: ProxyNode<Assume<TreeNodeSchema<Name, T>, StructSchema>, "javaScript">,
+	): SharedTreeObject<Assume<TreeNodeSchema<Name, T>, StructSchema>> {
+		const node = {};
```

Comment:
objects used like a map should be created with Object.create(null) to avoid inheriting members that may collide.
```suggestion
		const node = Object.create(null);
```

Reply:
Good call, thanks.